identity/aziot-identityd/src/lib.rs

```diff
@@ -47,105 +48,83 @@ macro_rules! match_id_type {
     };
 }
 
+pub enum ReprovisionTrigger {
+    ConfigurationFileUpdate,
+    Api,
+    Startup,
+}
+
 pub async fn main(
     settings: config::Settings,
+    config_path: std::path::PathBuf,
+    config_directory_path: std::path::PathBuf,
 ) -> Result<(http_common::Connector, http::Service), Box<dyn std::error::Error>> {
-    // Written to prev_settings_path if provisioning is successful.
     let settings = settings.check().map_err(InternalError::BadSettings)?;
-    let settings_serialized = toml::to_vec(&settings).expect("serializing settings cannot fail");
 
     let homedir_path = &settings.homedir;
     let connector = settings.endpoints.aziot_identityd.clone();
 
-    let mut prev_settings_path = homedir_path.clone();
-    prev_settings_path.push("prev_state");
-
-    let mut prev_device_info_path = homedir_path.clone();
-    prev_device_info_path.push("device_info");
-
     if !homedir_path.exists() {
         let () =
             std::fs::create_dir_all(&homedir_path).map_err(error::InternalError::CreateHomeDir)?;
     }
 
-    let (allowed_users, hub_modules, local_modules) =
-        prepare_authorized_principals(&settings.principal);
+    let api = Api::new(settings.clone())?;
+    let api = Arc::new(futures_util::lock::Mutex::new(api));
 
-    let allowed_users_copy = allowed_users.clone();
+    let api_startup = api.clone();
 
-    // All uids in the principals are authenticated users to this service
-    let authenticator = Box::new(move |uid| {
-        if allowed_users_copy.contains_key(&uid) {
-            Ok(auth::AuthId::LocalPrincipal(uid))
-        } else if uid.eq(&config::Uid(0)) {
-            Ok(auth::AuthId::LocalRoot)
-        } else {
-            Ok(auth::AuthId::Unknown)
-        }
-    });
+    let settings_copy = settings.clone();
 
-    let authorizer = Box::new(SettingsAuthorizer {
-        allowed_users: allowed_users.clone(),
-        modules: hub_modules.clone(),
+    tokio::task::spawn(async move {
+        let mut api_ = api_startup.lock().await;
+        let _ = api_
+            .update_settings(settings_copy, ReprovisionTrigger::Startup)
+            .await;
+        drop(api_);
     });
 
-    let api = Api::new(
-        settings,
-        local_modules,
-        authenticator,
-        authorizer,
-        allowed_users,
-    )?;
-    let api = Arc::new(futures_util::lock::Mutex::new(api));
+    let api_watcher = api.clone();
+    let (file_changed_tx, mut file_changed_rx) = tokio::sync::mpsc::channel(10);
```

Comment:
any significance to the number `10`, or did you just pick a number?
on that note, any reason this can't just be an `unbounded_channel` instead?

Reply:
IIRC the watcher can only work with a bounded one

Reply:
The watcher directly doesn't make any difference, but the point is that only the regular `Sender` has `blocking_send`. The `UnboundedSender` doesn't.

But anyway, I don't see a reason to use `10` either. `1` should be fine - there's only one sender and it's doing a blocking send, so it's fine if it blocks anyway.

Reply:
The main thread gets blocked if the message count is set to 1. Added a comment on why this value is being set to 2 for now. And I am certain we will need to tweak with this number at some point to optimize after testing.